src/hotspot/cpu/x86/c2_stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,1850 @@
+  broadcast_first_and_last_needle(needle, needle_len, rax, ae, _masm);
+
+  // Do "big switch" if haystack size > 32
+  __ cmpq(haystack_len, 0x20);
+  __ ja_b(L_bigSwitchTop);
+
+  // Copy the small (< 32 byte) haystack to the stack.  Allows for vector reads without page fault
+  // Only done for small haystacks
+  //
+  // NOTE: This code assumes that the haystack points to a java array type AND there are
+  //       at least 16 bytes of header preceeding the haystack pointer.
+  //
+  // This means that we're copying up to 15 bytes of the header onto the stack along
+  // with the haystack bytes.  After the copy completes, we adjust the haystack pointer
+  // to the valid haystack bytes on the stack.
+  {
+    Label L_moreThan16, L_adjustHaystack;
+
+    const Register index = rax;
+    const Register haystack = rbx;
+
+    // Only a single vector load/store of either 16 or 32 bytes
+    __ cmpq(haystack_len, 0x10);
+    __ ja_b(L_moreThan16);
+
+    __ movq(index, COPIED_HAYSTACK_STACK_OFFSET + 0x10);
+    __ movdqu(XMM_TMP1, Address(haystack, haystack_len, Address::times_1, -0x10));
+    __ movdqu(Address(rsp, COPIED_HAYSTACK_STACK_OFFSET), XMM_TMP1);
+    __ jmpb(L_adjustHaystack);
+
+    __ bind(L_moreThan16);
+    __ movq(index, COPIED_HAYSTACK_STACK_OFFSET + 0x20);
+    __ vmovdqu(XMM_TMP1, Address(haystack, haystack_len, Address::times_1, -0x20));
+    __ vmovdqu(Address(rsp, COPIED_HAYSTACK_STACK_OFFSET), XMM_TMP1);
+
+    // Point the haystack at the correct location of the first byte of the "real" haystack on the stack
+    __ bind(L_adjustHaystack);
+    __ subq(index, haystack_len);
+    __ leaq(haystack, Address(rsp, index, Address::times_1));
+  }
+
+  // Dispatch to handlers for small needle and small haystack
+  // Note that needle sizes of 1-6 have been handled in highly_optimized_short_cases,
+  // so the dispatch only has valid entries for 7-10.
+  __ leaq(r13, Address(save_ndl_len, -1));
+  __ cmpq(r13, NUMBER_OF_CASES - 1);
+  __ ja(L_smallCaseDefault);
+  __ lea(r15, InternalAddress(small_jump_table));
+  __ jmp(Address(r15, r13, Address::times_8));
+
+  // Dispatch to handlers for small needle and large haystack
+  // For large haystacks, the jump table is fully populated (1-10)
+  __ bind(L_bigSwitchTop);
+  __ leaq(rax, Address(save_ndl_len, -1));
+  __ cmpq(rax, NUMBER_OF_CASES - 1);
+  __ ja(L_bigCaseDefault);
+  __ lea(r15, InternalAddress(big_jump_table));
+  __ jmp(Address(r15, rax, Address::times_8));
+
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  // Fixup and return routines
+
+  // Return not found
+  __ bind(L_returnError);
+  __ movq(rax, -1);
+  __ jmpb(L_return);
+
+// Used to check and return value in rbp - usually error
+  __ bind(L_returnRBP);
+  __ movq(rax, rbp);
```

Comment:
This seems spurious as rax is being overwritten at line 489. Did you intend to return -1?

Reply:
Removed all references to L_returnRBP.  Replaced with L_returnError.